File: javascript/src/playback.test.ts
import * as path from 'path'
import { Vcr, VcrOutcome } from './index'

const tape = (name: string) => path.join(__dirname, '..', 'tapes', name)

describe('playback', () => {
  it('replays a recorded GET on a dynamic port', async () => {
    const vcr = Vcr.playback(tape('single_get.md'))
      .label('replays a recorded GET')
      .port(0)
      .start()
    try {
      expect(vcr.port).toBeGreaterThan(0)
      expect(vcr.tapeLength).toBe(1)

      const res = await fetch(`${vcr.baseUrl}/ok`)
      expect(res.status).toBe(200)
      expect(await res.text()).toBe('ok-body')
      expect(vcr.lastKind).toBe(VcrOutcome.Ok)
      expect(vcr.lastError).toBe('')
    } finally {
      vcr.close()
    }
  })

  it('flags a path mismatch via diagnostics', async () => {
    const vcr = Vcr.playback(tape('single_get.md')).port(0).start()
    try {
      await fetch(`${vcr.baseUrl}/nope`).catch(() => undefined)
      expect(vcr.lastKind).not.toBe(VcrOutcome.Ok)
      expect(vcr.lastError).not.toBe('')
    } finally {
      vcr.close()
    }
  })
})

File: dart/test/playback_test.dart
@TestOn('vm')
library;

import 'dart:io';

import 'package:servirtium/servirtium.dart';
import 'package:test/test.dart';

import 'support.dart';

String _tape(String name) => '${Directory.current.path}/test/tapes/$name';

void main() {
  test('replays a recorded GET on a dynamic port', () async {
    final vcr = Vcr.playback(_tape('single_get.md')).label('replays a GET').port(0).start();
    try {
      expect(vcr.port, greaterThan(0), reason: 'expected an OS-assigned port');
      expect(vcr.tapeLength, 1);

      final (status, body) = await httpGet('${vcr.baseUrl}/ok');

      expect(status, 200);
      expect(body, 'ok-body');
      expect(vcr.lastKind, VcrOutcome.ok);
      expect(vcr.lastError, isEmpty);
    } finally {
      vcr.close();
    }
  });

  test('flags a path mismatch via diagnostics', () async {
    final vcr = Vcr.playback(_tape('single_get.md')).port(0).start();
    try {
      await httpGet('${vcr.baseUrl}/nope');
      expect(vcr.lastKind, isNot(VcrOutcome.ok));
      expect(vcr.lastError, isNotEmpty);
    } finally {
      vcr.close();
    }
  });

  test('untaped path 404s without consuming the tape cursor', () async {
    final vcr = Vcr.playback(_tape('single_get.md')).untaped('/favicon.ico').port(0).start();
    try {
      // The incidental path is answered 404 and never touches the tape.
      final (status, _) = await httpGet('${vcr.baseUrl}/favicon.ico');
      expect(status, 404);

      // The recorded interaction still replays — the cursor wasn't consumed.
      final (okStatus, body) = await httpGet('${vcr.baseUrl}/ok');
      expect(okStatus, 200);
      expect(body, 'ok-body');
      expect(vcr.lastKind, VcrOutcome.ok);
    } finally {
      vcr.close();
    }
  });
}
